frontend/app/src/components/dialogs/WatcherDialog.vue

```diff
@@ -0,0 +1,380 @@
+      )!.args as WatcherArgs['makervault_collateralization_ratio'];
+      const modifiedWatcherArgs = watcher.args as WatcherArgs['makervault_collateralization_ratio'];
+
+      if (
+        existingWatcherArgs.op !== modifiedWatcherArgs.op ||
+        existingWatcherArgs.ratio !== modifiedWatcherArgs.ratio
+      ) {
+        let watcherPayload: Watcher[] = [];
+        watcherPayload.push(watcher);
+
+        const { commit } = this.$store;
+        this.$api
+          .editWatcher(watcherPayload)
+          .then(updatedWatchers => {
+            commit('balances/watchers', updatedWatchers);
+            commit('setMessage', {
+              title: 'Watcher Success',
+              description: 'Successfully edited the watcher.',
+              success: true
+            } as Message);
+            this.changeEditMode(watcher.identifier);
+            this.loadedWatchers = JSON.parse(JSON.stringify(updatedWatchers));
+          })
+          .catch((reason: Error) => {
+            commit('setMessage', {
+              title: 'Watcher Error',
+              description: reason.message || 'Error editing the watcher',
+              success: false
+            } as Message);
+          });
+      } else {
+        this.changeEditMode(watcher.identifier);
+      }
+    }
+  }
+
+  addWatcher() {
+    let watcherPayload: Omit<Watcher, 'identifier'>[] = [];
+
+    if (
+      this.watcherType &&
+      this.watcherValue &&
+      this.watcherOperation &&
+      this.watcherContentId
+    ) {
+      const { commit } = this.$store;
+      const watcherData: Omit<Watcher, 'identifier'> = {
+        type: this.watcherType as WatcherType,
+        args: {
+          ratio: this.watcherValue,
+          op: this.watcherOperation,
+          vault_id: this.watcherContentId.toString()
+        }
+      };
+      watcherPayload.push(watcherData);
+
+      this.$api
+        .addWatcher(watcherPayload)
+        .then(updatedWatchers => {
+          commit('balances/watchers', updatedWatchers);
+          commit('setMessage', {
+            title: 'Watcher Success',
+            description: 'Successfully added the watcher.',
+            success: true
+          } as Message);
+          this.watcherValue = null;
+          this.watcherOperation = null;
+          this.loadedWatchers = JSON.parse(JSON.stringify(updatedWatchers));
+        })
+        .catch((reason: Error) => {
+          commit('setMessage', {
+            title: 'Watcher Error',
+            description: reason.message || 'Error adding the watcher',
+            success: false
+          } as Message);
+        });
+    }
+  }
+
+  @Emit()
+  cancel() {
+    for (const index in this.existingWatchersEdit) {
+      // Reset edit mode on all fields
+      this.existingWatchersEdit[index] = false;
+
+      // Reset unsaved changes to the current saved state
+      this.loadedWatchers = JSON.parse(JSON.stringify(this.existingWatchers));
```

Comment:
again here copy by destructuring instead?

Reply:
I tried it and couldn't get it to work. I saw several places e.g. https://stackoverflow.com/questions/54801956/vue-changes-array-in-data-after-making-a-copy-of-it suggest this method works better for deep copies.